src/lib/optimizer/strategy/expression_reduction_rule.cpp

```diff
@@ -243,4 +248,114 @@ void ExpressionReductionRule::rewrite_like_prefix_wildcard(std::shared_ptr<Abstr
   }
 }
 
+void ExpressionReductionRule::remove_duplicate_aggregate(
+    std::vector<std::shared_ptr<AbstractExpression>>& input_expressions,
+    const std::shared_ptr<AbstractLQPNode>& aggregate_node, const std::shared_ptr<AbstractLQPNode>& root_node) {
+  // Create a list of all sums, counts, and averages in the aggregate node.
+  std::vector<std::reference_wrapper<const std::shared_ptr<AbstractExpression>>> sums, counts, avgs;
+  for (auto& input_expression : input_expressions) {
+    if (input_expression->type != ExpressionType::Aggregate) continue;
+    auto& aggregate_expression = static_cast<AggregateExpression&>(*input_expression);
+    switch (aggregate_expression.aggregate_function) {
+      case AggregateFunction::Sum: {
+        sums.emplace_back(input_expression);
+        break;
+      }
+      case AggregateFunction::Count: {
+        counts.emplace_back(input_expression);
+        break;
+      }
+      case AggregateFunction::Avg: {
+        avgs.emplace_back(input_expression);
+        break;
+      }
+      default:
+        continue;
+    }
+  }
+
+  // Take a copy of what the aggregate was originally supposed to do
+  const auto original_aggregate_expressions = aggregate_node->column_expressions();
+
+  const auto& aggregate_input_node = aggregate_node->left_input();
+  auto replacements = ExpressionUnorderedMap<std::shared_ptr<AbstractExpression>>{};
+
+  // Iterate over the AVGs, check if matching SUMs and COUNTs exist, and add a suitable replacement to `replacements`.
+  for (const auto& avg_expression_ptr : avgs) {
+    const auto& avg_expression = static_cast<AggregateExpression&>(*avg_expression_ptr.get());
+
+    const auto& avg_argument = avg_expression.argument();
+    const auto avg_argument_is_nullable = avg_argument.get()->is_nullable_on_lqp(*aggregate_input_node);
+
+    // A helper function that checks whether SUMs and COUNTs match the AVG
+    const auto finder = [&](const auto& other_expression) {
+      const auto other_argument = static_cast<const AggregateExpression*>(other_expression.get().get())->argument();
```

Comment:
There is probably an easy explanation that I don't see, but why are we using raw pointers here?